adapters/undertone/undertone.go

```diff
@@ -0,0 +1,214 @@
+		}
+		return nil, []error{err}
+	}
+
+	if responseData.StatusCode != http.StatusOK {
+		err := &errortypes.BadServerResponse{
+			Message: fmt.Sprintf("Unexpected status code: %d. Run with request.debug = 1 for more info.", responseData.StatusCode),
+		}
+		return nil, []error{err}
+	}
+
+	var response openrtb2.BidResponse
+	if err := json.Unmarshal(responseData.Body, &response); err != nil {
+		return nil, []error{err}
+	}
+
+	impIdBidTypeMap := map[string]openrtb_ext.BidType{}
+	for _, imp := range request.Imp {
+		if imp.Banner != nil {
+			impIdBidTypeMap[imp.ID] = openrtb_ext.BidTypeBanner
+		} else if imp.Video != nil {
+			impIdBidTypeMap[imp.ID] = openrtb_ext.BidTypeVideo
+		}
+	}
+
+	bidResponse := adapters.NewBidderResponseWithBidsCapacity(len(request.Imp))
+	bidResponse.Currency = response.Cur
+	for _, seatBid := range response.SeatBid {
+		for i, bid := range seatBid.Bid {
+			bidType, ok := impIdBidTypeMap[bid.ImpID]
+			if !ok {
+				continue
+			}
+
+			b := &adapters.TypedBid{
+				Bid:     &seatBid.Bid[i],
+				BidType: bidType,
+			}
+			bidResponse.Bids = append(bidResponse.Bids, b)
+		}
+		return bidResponse, nil
+	}
+	return bidResponse, nil
+}
+
+func (a *UndertoneAdapter) populateBidReqExt(bidRequest *openrtb2.BidRequest) {
+	undertoneBidderParams := &UndertoneBidderParams{
+		Id:      4,
+		Version: "1.0.0",
+	}
+	undertoneBidderParamsJSON, err := json.Marshal(undertoneBidderParams)
+	if err == nil {
+		extRequestPrebid := &openrtb_ext.ExtRequestPrebid{BidderParams: undertoneBidderParamsJSON}
+		bidRequestExt := &BidRequestExt{Prebid: extRequestPrebid}
+		bidRequestExtJSON, err2 := json.Marshal(bidRequestExt)
+		if err2 == nil {
+			bidRequest.Ext = bidRequestExtJSON
+		}
+	}
+}
+
+func (a *UndertoneAdapter) populateSiteApp(bidRequest *openrtb2.BidRequest, publisherId int, site *openrtb2.Site, app *openrtb2.App) {
+	pubId := strconv.Itoa(publisherId)
+	if site != nil {
+		siteCopy := *site
+		var publisher openrtb2.Publisher
+		if siteCopy.Publisher != nil {
+			publisher = *siteCopy.Publisher
+		}
+		publisher.ID = pubId
+		bidRequest.Site = &siteCopy
+		bidRequest.Site.Publisher = &publisher
+	} else if app != nil {
+		appCopy := *app
+		var publisher openrtb2.Publisher
+		if appCopy.Publisher != nil {
+			publisher = *appCopy.Publisher
+		}
+		publisher.ID = pubId
+		bidRequest.App = &appCopy
+		bidRequest.App.Publisher = &publisher
+	}
+}
+
+func (a *UndertoneAdapter) validateBidRequest(bidRequest *openrtb2.BidRequest) error {
+	if bidRequest.Site == nil && bidRequest.App == nil {
+		return fmt.Errorf("invalid req: missing Site/App")
+	}
+	return nil
+}
+
+func (a *UndertoneAdapter) parseImps(bidRequest *openrtb2.BidRequest) ([]openrtb2.Imp, int, []error) {
+	var errs []error
+	var publisherId int
+	var validImps []openrtb2.Imp
+
+	for _, imp := range bidRequest.Imp {
+		if imp.Banner == nil && imp.Video == nil {
```

Comment:
we do not need this check here because prebid core framework already checks that based on configuration defined in undertone.yaml file.